framework/mesh/mesh_continuum/mesh_continuum.cc

```diff
@@ -652,4 +653,88 @@ MeshContinuum::SetBoundaryIDFromLogical(const LogicalVolume& log_vol,
     grid_bndry_id_map[bndry_id] = boundary_name;
 }
 
+void
+MeshContinuum::ComputeVolumePerMaterialID() const
+{
+  auto grid_ptr = GetCurrentMesh();
+  const auto& grid = *grid_ptr;
+  std::shared_ptr<SpatialDiscretization> sdm_ptr = PieceWiseLinearContinuous::New(grid);
+  const auto& sdm = *sdm_ptr;
+
+  // Create a map to hold local volume with local material as key
+  std::map<int, double> material_volumes;
+  for (auto& cell : local_cells)
+  {
+    const auto& cell_mapping = sdm.GetCellMapping(cell);
+    const double volume = cell_mapping.CellVolume();
+    material_volumes[cell.material_id] += volume;
+  }
+
+  // Collect all local material IDs
+  std::set<int> unique_material_ids;
+  for (const auto& pair : material_volumes)
+    unique_material_ids.insert(pair.first);
+  // convert set to vector
+  std::vector<int> local_material_ids(unique_material_ids.begin(), unique_material_ids.end());
+  int local_size = local_material_ids.size();
+
+  // Initialize vector to hold sizes from all processes
+  std::vector<int> all_sizes(mpi_comm.size());
+  // Gather all local material ID sizes from all processes
+  mpi_comm.all_gather(local_size, all_sizes);
+
+  // Compute the displacement and total size
+  std::vector<int> displs(mpi_comm.size(), 0); // Initialize displacement vector
+  int total_size = 0;
+  for (int i = 0; i < mpi_comm.size(); ++i)
+  {
+    displs[i] = total_size;
+    total_size += all_sizes[i];
+  }
+
+  // Initialize vector to hold all material IDs from all processes
+  std::vector<int> all_material_ids(total_size);
+  // Gather all material IDs at root
+  mpi_comm.all_gather(local_material_ids, all_material_ids, all_sizes, displs);
+
+  // Create a union of all unique material IDs
+  std::set<int> global_unique_material_ids(all_material_ids.begin(), all_material_ids.end());
+
+  // Create vectors for the global reduction
+  std::vector<int> global_material_ids(global_unique_material_ids.begin(),
+                                       global_unique_material_ids.end());
+  std::vector<double> local_volumes(global_material_ids.size(), 0.0);
+  std::vector<double> global_volumes(global_material_ids.size(), 0.0);
+
+  // Fill local volumes vector based on the local material volumes
+  // and perform the reduction one material at a time
+  for (size_t i = 0; i < global_material_ids.size(); ++i)
+  {
+    if (material_volumes.find(global_material_ids[i]) != material_volumes.end())
+    {
+      local_volumes[i] = material_volumes[global_material_ids[i]];
+    }
+    double loc_vol = local_volumes[i];
+    double glo_vol;
+    mpi_comm.all_reduce(loc_vol, glo_vol, mpi::op::sum<double>());
+    global_volumes[i] = glo_vol;
+  }
+
+  // Combine results into a map on each process
+  std::map<int, double> global_material_volumes;
+  for (size_t i = 0; i < global_material_ids.size(); ++i)
+  {
+    global_material_volumes[global_material_ids[i]] = global_volumes[i];
+  }
```

Comment:
```suggestion
  // Fill local volumes vector based on the local material volumes
  // and perform the reduction one material at a time
  std::map<int, double> global_material_volumes;
  for (size_t i = 0; i < global_material_ids.size(); ++i)
  {
    if (material_volumes.find(global_material_ids[i]) != material_volumes.end())
      local_volumes[i] = material_volumes[global_material_ids[i]];
      
    mpi_comm.all_reduce(local_volumes[i], global_volumes[i], mpi::op::sum<double>());
    global_material_volumes[global_material_ids[i]] = global_volumes[i];
  }
```
Would this work?

Reply:
good question. I think the `mpi_comm.all_reduce` did not allow this when i coded it 5 months ago. Since then, David added that functionality. I will test it

Reply:
it works with the updated mpi communicator. thanks for the suggestion.